File: app/src/main/java/com/example/recipeclipper/ui/common/RecipeRow.kt
package com.example.recipeclipper.ui.common

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import coil.compose.AsyncImage
import com.example.recipeclipper.R
import com.example.recipeclipper.data.model.RecipeSummary

/** A recipe in a list: thumbnail, title, and when it was last viewed. */
@Composable
fun RecipeRow(recipe: RecipeSummary, now: Long, onClick: () -> Unit, modifier: Modifier = Modifier) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        modifier = modifier
            .fillMaxWidth()
            .clickable(role = Role.Button, onClick = onClick)
            .padding(vertical = 10.dp)
    ) {
        val shape = RoundedCornerShape(10.dp)
        if (recipe.imageUrl != null) {
            AsyncImage(
                model = recipe.imageUrl,
                contentDescription = null,
                contentScale = ContentScale.Crop,
                modifier = Modifier.size(64.dp).clip(shape)
            )
        } else {
            Box(Modifier.size(64.dp).clip(shape).background(MaterialTheme.colorScheme.outlineVariant))
        }
        Column(Modifier.weight(1f).padding(start = 14.dp)) {
            Text(
                recipe.title,
                style = MaterialTheme.typography.bodyLarge.copy(fontWeight = FontWeight.Bold),
                maxLines = 2,
                overflow = TextOverflow.Ellipsis
            )
            val details = listOfNotNull(
                if (recipe.isClipped) stringResource(R.string.clipped_by_you) else null,
                recipe.totalTime,
                TimeAgo.since(recipe.lastViewedAt, now).text()
            ).joinToString("  ·  ")
            Text(
                details,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
            if (recipe.isSaved) {
                Text(
                    stringResource(R.string.tag_saved),
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.tertiary
                )
            }
        }
    }
}

File: ios/RecipeClipper/UI/Common/RecipeRow.swift
import SwiftUI

/// A recipe in a list: thumbnail, title, total time and when it was last viewed, plus a
/// "Saved" tag when it is in any list.
struct RecipeRow: View {
    let recipe: RecipeSummary
    let now: Int64
    let onClick: () -> Void

    @Environment(\.dynamicTypeSize) private var dynamicTypeSize

    var body: some View {
        Button(action: onClick) {
            // At the accessibility sizes a thumbnail beside the text would leave it a column
            // a few letters wide, so the thumbnail goes above and the title may run longer.
            let stacked = dynamicTypeSize.isAccessibilitySize
            let layout = stacked
                ? AnyLayout(VStackLayout(alignment: .leading, spacing: 10))
                : AnyLayout(HStackLayout(spacing: 14))
            layout {
                Thumbnail(url: recipe.imageUrl)
                VStack(alignment: .leading, spacing: 2) {
                    Text(recipe.title)
                        .textStyle(Typography.bodyLargeBold)
                        .foregroundStyle(Palette.onBackground)
                        .lineLimit(stacked ? 5 : 2)
                        .multilineTextAlignment(.leading)
                    Text(details)
                        .textStyle(Typography.bodySmall)
                        .foregroundStyle(Palette.muted)
                    if recipe.isSaved {
                        Text(Strings.tagSaved)
                            .textStyle(Typography.labelSmall)
                            .foregroundStyle(Palette.accentText)
                    }
                }
                .frame(maxWidth: .infinity, alignment: .leading)
            }
            .padding(.vertical, 10)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }

    private var details: String {
        [recipe.isClipped ? Strings.clippedByYou : nil, recipe.totalTime, Strings.elapsed(TimeAgo.since(recipe.lastViewedAt, now: now))]
            .compactMap { $0 }
            .joined(separator: "  ·  ")
    }
}

private struct Thumbnail: View {
    let url: String?
    /// Grows a little with the text, capped: it is decoration, the title is the content.
    @ScaledMetric(relativeTo: .body) private var scaledSize: CGFloat = 64
    private var size: CGFloat { min(scaledSize, 96) }

    var body: some View {
        let shape = RoundedRectangle(cornerRadius: 10)
        ZStack {
            shape.fill(Palette.hairline)
            if let url, let parsed = URL(string: url) {
                CachedAsyncImage(url: parsed) { image in
                    image.resizable().scaledToFill()
                } placeholder: {
                    Color.clear
                }
            }
        }
        .frame(width: size, height: size)
        .clipShape(shape)
        .accessibilityHidden(true)
    }
}

/// The current wall-clock time in the same epoch-millisecond units the data layer uses.
/// Read once per screen appearance, like Android's `remember { System.currentTimeMillis() }`.
func currentMillis() -> Int64 { SystemClock().now() }
